### ISSUE TITLE
deployment ssh glitches from minimal config

### ISSUE THREAD

**Issue by tpapp** [JuliaDocs/Documenter.jl#625]
When using a minimal Docker image (based on Ubuntu 17.10) for deployment, I noticed the following minor bugs:

1. when `~/.ssh/` does not exist, [`withfile`](https://github.com/JuliaDocs/Documenter.jl/blob/5593b8f55f6ab191a7fdf76df867936a19d9e83a/src/Documenter.jl#L472) fails. **Workaround:** `mkdir ~/.ssh`.

2. when `github.com` is previously unknown to `ssh` (ie not in `known_hosts`), it would ask for confirmation in an interactive session, but [fetching upstream](https://github.com/JuliaDocs/Documenter.jl/blob/5593b8f55f6ab191a7fdf76df867936a19d9e83a/src/Documenter.jl#L488) fails in a non-interactive session. **Workaround**: populate a `~/.ssh/known_hosts` with the [`ssh` key fingerprints](https://help.github.com/articles/github-s-ssh-key-fingerprints/) for github (easiest: just run it interactively on an empty file and copy the result).

Again, most people would not notice these bugs because they have used `ssh` before on their machine to contact `github.com`. This is relevant for people running a minimal environment, like a virtual machine or a container. For this reason, I am not sure if it needs fixing explicitly, just maybe documenting the workarounds.

**Comment by mortenpi** [comment]
Thanks for writing these up!

1. I'd say it's a bug and we should just fix `withfile` here. Although, it is partially related to #458 -- i.e. we shouldn't really be touching user's SSH config in the first place.

2. In principle I think we could always run with SSH `StrictHostKeyChecking=no` to get around this.. but this does not feel like a good practice. The user should make sure that all the security-related stuff is set up properly. But we should mention it in the documentation for sure.

**Comment by tpapp** [comment]
Just to be clear: I was not suggesting abandoning key checking. I think the right workaround is what I outlined above (populate the `known_hosts`), and one cannot do any better. But with that, the other issue is solved, too (I think).

**Comment by ericphanson** [comment]
I've run into this too when trying to have the TagBot github actions workflow build stable docs (https://github.com/JuliaRegistries/TagBot/pull/49). My workaround was to add the known hosts in the workflow that runs documenter (see https://github.com/ericphanson/UnbalancedOptimalTransport.jl/blob/8585054cccb74a05a4624f1f915b9600f036eab2/.github/workflows/ci.yml#L40-L42). However, it was suggested it might be better if Documenter itself took care of that (https://github.com/invenia/PkgTemplates.jl/issues/130#issuecomment-576967591).

**Comment by mortenpi** [comment]
In #1216 we stopped modifying `.ssh/config`, so I am a little bit hesitant to go back to touching user's Git settings (even if it is just `known_hosts`). We could set [`StrictHostKeyChecking=accept-new`](https://man.openbsd.org/ssh_config#StrictHostKeyChecking) in the config though, which should get around the problem and still provide some security by failing when an existing key has changed?

**Comment by ericphanson** [comment]
Hmm, I am a bit confused— it seems like strict checking is off already? Or is this path not being hit?

https://github.com/JuliaDocs/Documenter.jl/blob/9c79330f92104d9a26793631f028d50b0ec85ace/src/Documenter.jl#L582

**Comment by mortenpi** [comment]
Oh, right. Any ideas why this setting appears to be ignored in your case? It should be exactly what we need?

> If this flag is set to “no” or “off”, ssh will automatically add new host keys to the user known hosts files and allow connections to hosts with changed hostkeys to proceed, subject to some restrictions.

Or are you running into the case where the IP has changed? We could also maybe add [`CheckHostIP=no`](https://man.openbsd.org/ssh_config#CheckHostIP).

**Comment by ericphanson** [comment]
This was the specific error: https://github.com/ericphanson/UnbalancedOptimalTransport.jl/runs/395589862#step:5:79

**Comment by mortenpi** [comment]
It could be that your `~/.ssh/known_hosts` has some weird permissions? If so, one workaround would be to use our own `known_hosts` file by setting [`UserKnownHostsFile`](https://man.openbsd.org/ssh_config#UserKnownHostsFile) to a temporary file.

**Comment by christopher-dG** [comment]
I'm pretty sure this is all happening in a temporary CI environment, so modifying settings shouldn't really be a problem. IMO the `UserKnownHostsFile` is a good plan (that's what TagBot is about to switch to).

**Comment by fredrikekre** [comment]
Did the message `Failed to add the RSA host key for IP address '192.30.253.113' to the list of known hosts (/home/runner/.ssh/known_hosts).` really block the push? I think that for #1216 I noticed this warning too, and I even fixed it, but the workaround was pretty awkward so I skipped it and hoped noone would see it. But as I remember it it was not actually blocking the push, it was just a warning. Maybe it was a different message though...

**Comment by ericphanson** [comment]
I think it did; at least, I did not have stable docs after that run, but I did once it ran with https://github.com/ericphanson/UnbalancedOptimalTransport.jl/blob/8585054cccb74a05a4624f1f915b9600f036eab2/.github/workflows/ci.yml#L40-L42.

**Comment by mortenpi** [comment]
@ericphanson isn't [this `gh-pages` commit](https://github.com/ericphanson/UnbalancedOptimalTransport.jl/commit/8e576bd13a8268583aaf3297f5d13e466eba9300) from the [GitHub actions run that showed the key adding error](https://github.com/ericphanson/UnbalancedOptimalTransport.jl/runs/395589862#step:5:79)? The commit IDs appear to match at least.

**Comment by ericphanson** [comment]
> @ericphanson isn't [this `gh-pages` commit](https://github.com/ericphanson/UnbalancedOptimalTransport.jl/commit/8e576bd13a8268583aaf3297f5d13e466eba9300) from the [GitHub actions run that showed the key adding error](https://github.com/ericphanson/UnbalancedOptimalTransport.jl/runs/395589862#step:5:79)? The commit IDs appear to match at least.

You're right, looks like it did push successfully. I'm not sure what went wrong when I tried it at the time; maybe I tried right after the action finished but Github pages still needed to rerun jekyll or something?
